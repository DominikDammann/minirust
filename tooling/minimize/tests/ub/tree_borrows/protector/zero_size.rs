//@ compile-flags: --minimize-tree-borrows

// Check that protectors work even with a zero-sized pointee.

fn foo(x: *mut i32, y: &mut()) {
    // (x, Reserved) -> (y, P[Reserved])
    let y = y as *mut () as *mut i32;
    unsafe { *y = 42 }; // y becomes Accessed and Active.
    unsafe { assert!(*x == 42) }; // UB! Foreign Write to Protected Active.
    
}

fn main() {
    let x = &mut 31; // (x, Reserved)   
    let xraw = x as *mut i32; // (x, Reserved)
    let y = unsafe { &mut *(xraw as *mut ()) }; // (x, Reserved) -> (y, Reserved)

    foo(xraw, y);
}
