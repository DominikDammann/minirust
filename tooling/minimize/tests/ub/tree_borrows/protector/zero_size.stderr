fatal error: UB: Tree Borrows: a protected pointer with Unique permission becomes Disabled
